### ISSUE TITLE
Error: IPC server timed out before receiving a client connection

### ISSUE THREAD

**Issue by j-windsor** [hbenl/mocha-explorer-launcher-scripts#1]
I am getting this error when trying to refresh tests in the Mocha Explorer in your [helloworld-test-sample](https://github.com/hbenl/vscode-extension-samples/tree/test-explorer-integration/helloworld-test-sample)

```
Couldn't establish IPC:
Error: IPC server timed out before receiving a client connection
	at Timeout.onTimeout [as _onTimeout] (/Users/jaywindsor/.vscode/extensions/hbenl.vscode-mocha-test-adapter-2.5.0/node_modules/vscode-test-adapter-remoting-util/out/ipc.js:95:20)
	at listOnTimeout (internal/timers.js:531:17)
	at processTimers (internal/timers.js:475:7)
```

I cloned the repo, followed the readme steps, and this is where I am blocked. Any tips?

**Comment by hbenl** [comment]
This is probably because when you refresh the tests for the first time, `vscode-test` will download a current version of VS Code Insiders and unpack it, which usually takes some time, so the launcher script runs into a timeout waiting for a reply from the tests that will run in VS Code Insiders. Wait for a minute and then try again. If this doesn't solve the problem, set `mochaExplorer.logpanel` to `true` in your settings and have a look at the "Mocha Explorer Log" output channel, it will contain more detailed information. Paste the output here if you can't figure out why it doesn't work.

**Comment by j-windsor** [comment]
Thanks @hbenl! Looks like it is trying to create a directory in the root directory that it does not have access to.

```[2020-06-19 18:29:02.410] [INFO] Test Explorer found
[2020-06-19 18:29:02.411] [INFO] Creating adapter for /Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample
[2020-06-19 18:29:02.411] [INFO] Registering adapter for /Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample
[2020-06-19 18:29:02.412] [INFO] Skipping the initial load request for /Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample
[2020-06-19 18:29:02.414] [INFO] Initialization finished
[2020-06-19 18:29:12.673] [INFO] Loading test files of /Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample
[2020-06-19 18:29:12.674] [DEBUG] Using working directory: /Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample
[2020-06-19 18:29:13.092] [DEBUG] Using Mocha options: {"ui":"tdd","timeout":2000,"retries":0,"requires":[],"delay":false,"fullTrace":false,"exit":false}
[2020-06-19 18:29:13.092] [DEBUG] Looking for test files ["out/test/suite/*.test.js"] in /Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample
[2020-06-19 18:29:13.143] [DEBUG] Found test files ["/Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample/out/test/suite/extension.test.js"]
[2020-06-19 18:29:13.164] [DEBUG] Using nodePath: /usr/local/bin/node
[2020-06-19 18:29:13.164] [DEBUG] Using environment variables from config: {"VSCODE_VERSION":"insiders","ELECTRON_RUN_AS_NODE":null}
[2020-06-19 18:29:13.166] [DEBUG] Spawning /Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample/node_modules/mocha-explorer-launcher-scripts/vscode-test with IPC options {"role":"client","port":9449,"host":"localhost"}
[2020-06-19 18:29:13.481] [ERROR] Worker (stderr): Error: EROFS: read-only file system, mkdir '/.vscode-test'
    at Object.mkdirSync (fs.js:841:3)
    at /Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample/node_modules/vscode-test/out/download.js:92:20
    at step (/Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample/node_modules/vscode-test/out/download.js:36:23)
    at Object.next (/Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample/node_modules/vscode-test/out/download.js:17:53)
    at /Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample/node_modules/vscode-test/out/download.js:11:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample/node_modules/vscode-test/out/download.js:7:12)
    at downloadVSCodeArchive (/Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample/node_modules/vscode-test/out/download.js:89:12)
    at Object.<anonymous> (/Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample/node_modules/vscode-test/out/download.js:242:42)
    at step (/Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample/node_modules/vscode-test/out/download.js:36:23) {
  errno: -30,
  syscall: 'mkdir',
  code: 'EROFS',
  path: '/.vscode-test'
}

[2020-06-19 18:29:13.481] [ERROR] Worker (stderr): Failed to run tests: Error: Failed to download and unzip VS Code insiders
    at Object.<anonymous> (/Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample/node_modules/vscode-test/out/download.js:255:27)
    at step (/Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample/node_modules/vscode-test/out/download.js:36:23)
    at Object.throw (/Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample/node_modules/vscode-test/out/download.js:17:53)
    at rejected (/Users/jaywindsor/Downloads/vscode-extension-samples-test-explorer-integration/helloworld-test-sample/node_modules/vscode-test/out/download.js:9:65)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)

[2020-06-19 18:29:18.179] [ERROR] Couldn't establish IPC: [Error: IPC server timed out before receiving a client connection
	at Timeout.onTimeout [as _onTimeout] (/Users/jaywindsor/.vscode/extensions/hbenl.vscode-mocha-test-adapter-2.5.0/node_modules/vscode-test-adapter-remoting-util/out/ipc.js:95:20)
	at listOnTimeout (internal/timers.js:531:17)
	at processTimers (internal/timers.js:475:7)]

**Comment by hbenl** [comment]
Damn, I remember I saw the same error once but I don't remember what the reason for it was.
What version of VS Code do you use? If you're using the "Insiders" version, you'll have to change `"VSCODE_VERSION": "insiders"` to `"VSCODE_VERSION": "stable"` in `.vscode/settings.json`, because the version used for running the tests must be different than the version in which the project is open.

**Comment by j-windsor** [comment]
I'm on v1.46 stable and have `"VSCODE_VERSION": "insiders"` set.

**Comment by hbenl** [comment]
OK, that's correct.
I'll do some experiments, hopefully I'll be able to reproduce this (again).

**Comment by j-windsor** [comment]
Great! Thanks for the help! I do think it is related to the permissions to write and read that folder. I run VS Code as su and it works.

**Comment by hbenl** [comment]
I remembered [this](https://github.com/microsoft/vscode-test/issues/57#issuecomment-597209123) and so I could reproduce and fix it.
As it turns out, VS Code on Mac runs extensions using `/` as the working directory, and `vscode-test` tries to create the directory `.vscode-test` in the working directory. I fixed this in version 2.6.0: child processes are now started in the correct working directory (usually the workspace).
Furthermore I have added the config option `mochaExplorer.ipcTimeout` so you can increase the timeout (but this should only be necessary if your Mac is quite slow).
Please tell me if it works for you now.

**Comment by redborian** [comment]
I increased the timeout and was able to get it working. Thanks @hbenl!
